### PR TITLE
 planner, executor: tiny refactor

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Let codes in executor don't rely on `expression.Column.DBName` and `expression.Column.TblName`

And some refactor to reuse some codes and remove some duplicated codes.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: executor/delete.go [hunk 1]
```diff
@@ -132,50 +114,20 @@ func (e *DeleteExec) deleteSingleTableByChunk(ctx context.Context) error {
...
-		// tblRowMap[info.tblID][handle] hold the row datas binding to this table and this handle.
-		tblRowMap[info.tblID][handle] = joinedRow[info.colBeginIndex:info.colEndIndex]
+		handle := joinedRow[info.HandleOrdinal].GetInt64()
+		// tblRowMap[info.TblID][handle] hold the row datas binding to this table and this handle.
+		tblRowMap[info.TblID][handle] = joinedRow[info.Start:info.End]
 	}
 }
 
 func (e *DeleteExec) deleteMultiTablesByChunk(ctx context.Context) error {
-	if len(e.Tables) == 0 {
```
Comment (winoros) [inline executor/delete.go]: I checked the grammer in parser. If it's a multi delete. The `Tables` cannot be nil.

Comment (jackysp) [reply executor/delete.go]: OK

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -2522,6 +2522,77 @@ func (b *PlanBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
 	return joinPlan, nil
 }
 
+func getTableOffset(schema *expression.Schema, handleCol *expression.Column) int {
+	for i, col := range schema.Columns {
+		if col.DBName.L == handleCol.DBName.L && col.TblName.L == handleCol.TblName.L {
+			return i
+		}
+	}
+	panic("Couldn't get column information when do update/delete")
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: it's better to return an error rather than just panic.

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2522,6 +2522,77 @@ func (b *PlanBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
...
+}
+
+// Less implements sort.Interface#Less.
+func (c TblColPosInfoSlice) Less(i, j int) bool {
+	return c[i].Start < c[j].Start
+}
+
+// FindHandle finds the ordinal of the corresponding handle column.
+func (c TblColPosInfoSlice) FindHandle(ordinal int) (int, bool) {
+	if c == nil || len(c) == 0 {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: `len(c) == 0` can cover the case of `c == nil`.

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -2522,6 +2522,77 @@ func (b *PlanBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
...
+	c[i], c[j] = c[j], c[i]
+}
+
+// Less implements sort.Interface#Less.
+func (c TblColPosInfoSlice) Less(i, j int) bool {
+	return c[i].Start < c[j].Start
+}
+
+// FindHandle finds the ordinal of the corresponding handle column.
+func (c TblColPosInfoSlice) FindHandle(ordinal int) (int, bool) {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: 1. Should the slice be sorted before calling this function?
2. How about s/ordinal/colOrdinal/?

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -2522,6 +2522,77 @@ func (b *PlanBuilder) buildSemiJoin(outerPlan, innerPlan LogicalPlan, onConditio
...
+		tbl := tblID2Table[tblID]
+		var tblLen int
+		if onlyWritableCol {
+			tblLen = len(tbl.WritableCols())
+		} else {
+			tblLen = len(tbl.Cols())
+		}
+		for _, handleCol := range handleCols {
+			offset := getTableOffset(schema, handleCol)
+			end := offset + tblLen
```
Comment (qw4990) [inline planner/core/logical_plan_builder.go]: I have two questions about this 🤔:
1. Can we guarantee that columns of the same table are adjacent in `Schema`?
2. This `tblLen` is correct after column pruning?

Comment (winoros) [reply planner/core/logical_plan_builder.go]: For 1, yes, we keep this assumption from the very beginning.
For 2, no column will be pruned when updating/deleting. You can see the orignal codes. In executing phase, it still use `tbl.WritableCols()` or `tbl.Cols()` as the length.

Comment (qw4990) [reply planner/core/logical_plan_builder.go]: Okay.

File: planner/core/logical_plan_builder.go (hunk 4 above)
Comment (winoros) [reply planner/core/logical_plan_builder.go]: It's sorted when building it.

**Pull request merged**